src/main/scala/it/pagopa/interop/agreementmanagement/api/impl/AgreementApiServiceImpl.scala

```diff
@@ -279,83 +282,139 @@ final case class AgreementApiServiceImpl(
+        )
+      )
 
     onComplete(result) {
-      case Success(statusReply) if statusReply.isSuccess =>
-        updateAgreementVerifiedAttribute200(PersistentAgreement.toAPI(statusReply.getValue))
-      case Success(statusReply)                          =>
-        logger.error(
-          s"Error while $operationLabel $agreementId verified attribute ${verifiedAttributeSeed.id}",
-          statusReply.getError
-        )
-        statusReply.getError match {
-          case ex: AgreementNotFound => updateAgreementVerifiedAttribute404(problemOf(StatusCodes.NotFound, ex))
-          case ex                    => internalServerError(operationLabel, agreementId, ex.getMessage)
-        }
-      case Failure(ex)                                   =>
-        logger.error(s"Error while $operationLabel $agreementId verified attribute ${verifiedAttributeSeed.id}", ex)
+      case Success(document)              =>
+        addAgreementConsumerDocument200(PersistentAgreementDocument.toAPI(document))
+      case Failure(ex: AgreementNotFound) =>
+        logger.error(s"Error while $operationLabel", ex)
+        addAgreementConsumerDocument404(problemOf(StatusCodes.NotFound, ex))
+      case Failure(ex)                    =>
+        logger.error(s"Error while $operationLabel", ex)
+        internalServerError(operationLabel, agreementId, ex.getMessage)
+    }
+  }
+
+  override def removeAgreementConsumerDocument(agreementId: String, documentId: String)(implicit
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem],
+    contexts: Seq[(String, String)]
+  ): Route = authorize(ADMIN_ROLE) {
+
+    val operationLabel: String = s"Removing consumer document $documentId from agreement $agreementId"
+
+    logger.info(operationLabel)
+
+    val commander: EntityRef[Command] =
+      sharding.entityRefFor(AgreementPersistentBehavior.TypeKey, getShard(agreementId))
+
+    val result: Future[Unit] =
+      commander.askWithStatus(ref => RemoveAgreementConsumerDocument(agreementId, documentId, ref))
+
+    onComplete(result) {
+      case Success(_)                     =>
+        removeAgreementConsumerDocument204
+      case Failure(ex: AgreementNotFound) =>
+        logger.error(s"Error while $operationLabel", ex)
+        removeAgreementConsumerDocument404(problemOf(StatusCodes.NotFound, ex))
+      case Failure(ex)                    =>
+        logger.error(s"Error while $operationLabel", ex)
+        internalServerError(operationLabel, agreementId, ex.getMessage)
+    }
+  }
+
+  override def getAgreementConsumerDocument(agreementId: String, documentId: String)(implicit
+    toEntityMarshallerDocument: ToEntityMarshaller[Document],
+    toEntityMarshallerProblem: ToEntityMarshaller[Problem],
+    contexts: Seq[(String, String)]
+  ): Route = authorize(ADMIN_ROLE) {
+
+    val operationLabel: String = s"Retrieving consumer document $documentId from agreement $agreementId"
+
+    logger.info(operationLabel)
+
+    val commander: EntityRef[Command] =
```

Comment:
Refactor suggestion: most of the time, you must get the shard according to some kind of String id. Why don't you create something like

```scala
private def commander(id: String): EntityRef[Command] =
    sharding.entityRefFor(AgreementPersistentBehavior.TypeKey, getShard(id))
```

to avoid one assignation in every endpoint. The final result could be like

```scala
val result: Future[PersistentAgreementDocument] = commander(agreementId)
    .askWithStatus(ref => GetAgreementConsumerDocument(agreementId, documentId, ref))
```

Reply:
much better 👌